### DISCUSSION TITLE
Combination of multiple hosts and tasks

### DISCUSSION THREAD

**Discussion by MichalObs97** [nornir-automation/nornir#734] [Q&A]
Hi, 

I have wrote a script for upgrading a image version for Cisco WLCs, that follows a series of steps during the whole process of upgrade. I structured it so the task is written in its own function, and after execution, next function is called. Example: 
 
    def save_config(task): 
        task.run(task = netmiko_send_command, command_string = 'save config', expect_string=r'Are you sure you want to save')
        task.run(task = netmiko_send_command, command_string = 'y', delay_factor=2, expect_string=r'Configuration Saved')
        print('Configuration Saved, Proceeding to Reboot')
        reboot (task)

    def reboot(task):
        time.sleep(5)
        task.run(task = netmiko_send_command, command_string = 'reset system both', expect_string=r'Are you sure')
        task.run(task = netmiko_send_command, command_string = 'y', expect_string=r'System will now restart')
        print('Restart in progress')
   
Problem happens whenever I upload a image to a two different hosts with two different speeds and therefore one is much faster than the other one. Since I have to reboot the device to boot up a new image, I lose the connection so I close all connections and connect again. At that point some devices are waiting a long time for other to get to the exactly same point, so all connections could be closed at the same time and connected again. 

My question would be: Is there a way somehow to call a next task on one device while second device is still on the first task? I am looking for something other than calling a next task from the task before, because at some point they will wait for each other anyways. Ideal situation would be that the faster device would be done with all of the tasks while the slower one is still working on something. 

(I am calling the main tasks from main function for example the whole upload process all the way to restart is called by one task in main and then tasks call next tasks and so on all the way to the restart. As far as I understand it, there is only one instance of main function running, but each task function runs x times depending on amount of hosts - x.) 

Thank you so much!

**Comment by ktbyers** [comment]
@MichalObs97 Can you step back and describe how you want to it work? Take a case with two WLCs one slow and one fast to simplify the problem.

**Reply by MichalObs97** [reply]
@ktbyers Yes of course! Lets say I have two controllers as you said. At the beginning I have to run tests if the controller is on the "desired" image version or not, and check the redundancy information. If controller fails at least one task, the process is terminated on that particular controller. If controller passes these two tests it continues forward with the process -- checks how many APs are registered to the controller and then image upload starts. Lets say so far controllers would be in the similar spot. Images upload and install, then controller saves the config and reboots. After reboot I am checking if the port 22 is back up so I can connect to the controller again. At that point I need to close any existing connection and connect to the controller again to carry on with a verification process. But at that point fast controller could be rebooted and back up, while slow controller is still just on the uploading image task (task that faster one finished X minutes ago). Since I am calling the "main" tasks from the main function I have to wait for both controllers to be up after reboot to continue to verification process. 

    def main():
        nr = InitNornir('network_automation/aireos_code_upgrade/inventory/config.yaml', runner={"plugin": "threaded", "options": {"num_workers": 30}})
        global result, AP_before
        nr.run(task = paging)
        result = nr.run(task = check_redundancy)
        print_result(result)
        result1 = nr.run(task = check_image)
        print_result(result1)
        nr.run(task = after_initial_check)
        words=readFile('filtered_hosts.txt')
        nr.close_connections(nr.inventory.hosts)
        nr = InitNornir('network_automation/aireos_code_upgrade/inventory/config.yaml', runner={"plugin": "threaded", "options:{"num_workers": 30}})
        runbook_nr_filter = nr.filter(F(hostname__any=words)) 
        runbook_nr_filter.run(task = updatecheck)
        result2 = runbook_nr_filter.run(task = terminal_message)
        print_result(result2)

My question would be, if there is a way to modify my script so faster controller do not have to wait for slower one to finish the "after_initial_check" task chain described in the earlier message. I would want faster controller to finish the whole script on its own while slower one could still be somewhere in the process. Thank you so much for your help!

**Comment by dbarrosop** [comment]
Create a task with both tasks as subtasks and the second subtask will be called in parallel with the first one.

For reference; https://nornir.readthedocs.io/en/latest/plugins/execution_model.html

**Reply by MichalObs97** [reply]
Thank you so much for your solution @dbarrosop! I used something similar to what you are recommending me by connecting one task to the other one, creating a "chain-like" tasks. But my problem is the point where script has to go trough the disconnect and connect again cause that particular part only worked for me when it was in the main function. Thank you so much for your help!
